Add warnings related to Microsoft.NET.Sdk.WindowsDesktop

vatsan-madhavan opened the pull request:
Related: #3126 
This is also needed to completely fix https://github.com/dotnet/wpf/issues/866 - we still can't support support multitargeting `netcoreapp2.2`, and also to fix https://github.com/dotnet/wpf/issues/867

- Adds warning strings related to `Microsoft.NET.Sdk.WindowsDesktop`
- Adds a check to warn when `UseWpf` or `UseWindowsForms` is set without including `Microsoft.NET.Sdk.WindowsDesktop`

vatsan-madhavan commented:
/cc @rladuca

peterhuene commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -538,4 +538,16 @@ The following are names of parameters or literal values and should not be transl
+    <comment>{StrBegin="NETSDK1105: "}</comment>
+  </data>
+  <data name="UseWpfOrUseWindowsFormsRequiresWindowsDesktopFramework" xml:space="preserve">
+    <value>NETSDK1107: Microsoft.NET.Sdk.WindowsDesktop is required to build Windows desktop applications. 'UseWpf' and 'UseWindowsForms' are not supported by the current Sdk.</value>
```
Nit:

```suggestion
    <value>NETSDK1107: Microsoft.NET.Sdk.WindowsDesktop is required to build Windows desktop applications. 'UseWpf' and 'UseWindowsForms' are not supported by the current SDK.</value>
```

peterhuene commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -538,4 +538,16 @@ The following are names of parameters or literal values and should not be transl
     <value>NETSDK1104: RollForward value '{0}' is invalid. Allowed values are {1}.</value>
     <comment>{StrBegin="NETSDK1104: "}</comment>
   </data>
+  <data name="WindowsDesktopFrameworkRequiresVersion30" xml:space="preserve">
```
Is this going to be used by a future change?

vatsan-madhavan replied:
Yes, as soon as this change flows into `dotnet/wpf`, we will consume this in `Microsoft.NET.Sdk.WindowsDesktop`

peterhuene commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -538,4 +538,16 @@ The following are names of parameters or literal values and should not be transl
+    <value>NETSDK1107: Microsoft.NET.Sdk.WindowsDesktop is required to build Windows desktop applications. 'UseWpf' and 'UseWindowsForms' are not supported by the current Sdk.</value>
+    <comment>{StrBegin="NETSDK1107: "}</comment>
+  </data>
+  <data name="WindowsDesktopFrameworkRequiresUseWpfOrUseWindowsForms" xml:space="preserve">
```
Is this going to be used by a future change?

vatsan-madhavan replied:
Yes, as soon as this change flows into `dotnet/wpf`, we will consume this in `Microsoft.NET.Sdk.WindowsDesktop`

peterhuene reviewed (approved):
Looks good, just a few questions and a nit.

peterhuene commented:
Tests are failing with the new warning (treated as error).  Perhaps there are some test scenarios that need updating?

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.targets:
```diff
@@ -333,4 +333,11 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <Target Name="CheckWindowsDesktopSdkInUse" 
+          BeforeTargets="_CheckForInvalidConfigurationAndPlatform" 
+          Condition="'$(UseWpf)' == 'true' Or '$(UseWindowsForms)' == 'true'">
+    <NetSdkWarning Condition="'$(_MicrosoftNetSdkWindowsDesktop)' != 'true'" 
```
It looks like `_MicrosoftNetSdkWindowsDesktop` isn't set by the WindowsDesktop SDK yet.  Until it is, this will block WindowsDesktop projects from building.  I think you need to set that property in the WindowsDesktop SDK first, flow that through to core-sdk and then to stage 0 in this repo, before committing this change.

vatsan-madhavan replied:
I'm watching DARC flows (I triggered one earlier this morning from core-setup -> toolset exactly for this purpose: https://github.com/dotnet/toolset/pull/1177). Won't merge this until it's _very safe_ to do so.

vatsan-madhavan replied:
Update global.json with latest core-sdk that contains support for `_MicrosoftNetSdkWindowsDesktop`

nguerrera commented on src/Tasks/Common/Resources/xlf/Strings.pt-BR.xlf:
```diff
@@ -1,80 +1,80 @@
       <trans-unit id="AddResourceWithNonIntegerResource">
         <source>NETSDK1076: AddResource can only be used with integer resource types.</source>
-        <target state="translated">NETSDK1076: AddResource só pode ser usado com tipos de recursos inteiros.</target>
+        <target state="new">NETSDK1076: AddResource can only be used with integer resource types.</target>
```
Something went wrong and things are resetting to English in some files. Can you try reverting all the xlf and let it regenerate? Also, if you have any insight on how this happened, let's add an issue to github.com/dotnet/xliff-tasks.

peterhuene replied:
Nice catch as I missed this (unexpanded) file.  I wonder what caused this.

peterhuene replied:
It happened to `src/Tasks/Common/Resources/xlf/Strings.tr.xlf` too.

vatsan-madhavan replied:
Looking through the history, this happened in the very first commit. I'll redo the work and look through each file this time... and see if I can repro the problem along the way..

vatsan-madhavan replied:
Fixed these.

vatsan-madhavan committed:
- Add warning strings related to Microsoft.NET.Sdk.WindowsDesktop
- Add a check to warn when `UseWpf` or `UseWindowsForms` is set without including `Microsoft.NET.Sdk.WindowsDesktop`

vatsan-madhavan committed:
- Add warning strings related to Microsoft.NET.Sdk.WindowsDesktop
- Add a check to warn when `UseWpf` or `UseWindowsForms` is set without including `Microsoft.NET.Sdk.WindowsDesktop`

vatsan-madhavan commented:
Anyone have ideas on why `Windows_NT_TestAsTools Build_Debug`  is failing ?

dsplaisted commented:
@wli3 Can you look at the TestAsTools failure here?  It looks like an issue with local tools and/or Json parsing.  (It's possible that the way the tool is being run is mixing up different pieces of the .NET Core SDK from stage 0 and from the repo in an incompatible way after stage 0 has been updated).

```
    System.TypeLoadException: Could not load type 'System.Text.Json.Serialization.JsonSerializer' from assembly 'System.Text.Json, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.
       at Microsoft.DotNet.ToolPackage.LocalToolsResolverCache.Save(IDictionary`2 restoredCommandMap)
       at Microsoft.DotNet.Tools.Tool.Install.LocalToolsResolverCacheExtensions.SaveToolPackage(ILocalToolsResolverCache localToolsResolverCache, IToolPackage toolDownloadedPackage, String targetFrameworkToInstall)
       at Microsoft.DotNet.Tools.Tool.Install.ToolInstallLocalCommand.Install(FilePath manifestFile)
       at Microsoft.DotNet.Tools.Tool.Install.ToolInstallLocalCommand.Execute()
       at Microsoft.DotNet.Tools.Tool.Install.ToolInstallCommand.Execute()
       at Microsoft.DotNet.Cli.DotNetTopLevelCommandBase.RunCommand(String[] args)
       at Microsoft.DotNet.Tools.Tool.ToolCommand.Run(String[] args)
       at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)
       at Microsoft.DotNet.Cli.Program.Main(String[] args)
    Errors
        F:\workspace\_work\1\s\src\Tests\Directory.Build.targets(56,5): error MSB3073: The command "dotnet tool install --local testSdkClean --version 3.0.100-ci --add-source F:\workspace\_work\1\s\artifacts\packages\Debug\NonShipping\" exited with code 1. [F:\workspace\_work\1\s\src\Tests\Microsoft.NET.Clean.Tests\Microsoft.NET.Clean.Tests.csproj]
```

wli3 commented:
@dsplaisted yes. there was a breaking change from s.t.json. and we get a mix of assembly versions.

wli3 commented:
update stage 0 should resolve it

wli3 commented on global.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "3.0.100-preview7-012270",
+    "dotnet": "3.0.100-preview7-012333",
```
@vatsan-madhavan I cannot push to this branch. Could you update this line to `3.0.100-preview7-012358`?

wli3 commented:
never mind, there should be a new breaking change. I need to push it though core-sdk to fix it

wli3 commented:
Here is the breaking change https://github.com/dotnet/corefx/pull/38120 cc @ahsonkhan

wli3 commented:
my PR is here https://github.com/dotnet/cli/pull/11532 however, after merging the change need to flow to core-sdk. and then we need to update the stage 0 in this PR to the new build core-sdk

wli3 commented:
@dsplaisted @livarcocc we need to wait for about a day to check this in. or we can consider ignore "Windows_NT_TestAsTools Build_Debug" if we need to get it in quick

vatsan-madhavan committed:
Update Stage 0

vatsan-madhavan commented:
@wli3 Windows_NT_TestAsTools Build_Debug is continuing to fail with the latest core-sdk in Stage 0. Any ideas when this will be fixed? I'm eager to get this change to get in since there is some preview 7 work in WPF that depends on these changes flowing through to core-sdk and then onto dotnet/wpf....

wli3 commented:
We have an CI issue that is blocking the dependency flow. I will add you to the thread

https://github.com/dotnet/toolset/pull/1207

wli3 commented:
@vatsan-madhavan meanwhile do you mind add me to your fork as contributor? I can push it through finish if i have the access to push to the branch

nguerrera commented:
Normally you have push access to the PR source branch if PR has the checkbox to allow maintainers to do so (which it does by default). This allows this without maintaining collaborator lists in forks.

vatsan-madhavan commented:
@wli3 just added you. if i unchecked the 'allow maintainers to edit..' checkbox, i have no memory of doing so. when i try to create another PR from my fork, the checkbox is definitely enabled by default as @nguerrera describes it (which means I must have unchecked it :( and forgotten all about it... ).

dsplaisted committed:
Don't hash DesignTimeBould for ResolvePackageAssets

Commit:
Update stage 0

Commit:
Adjust tests for runtime pack name change

wli3 commented:
also added e6c78c75da7556b916da5d010aa0d976a46e6b19 to adjust runtime pack name change talked to @dsplaisted

wli3 commented:
@fadimounir https://dev.azure.com/dnceng/public/_build/results?buildId=224313&view=ms.vss-test-web.build-test-results-tab these new failing tests are in your area. Do you know them?

fadimounir commented:
Looks like some update that is maybe breaking the logic around this line: https://github.com/dotnet/sdk/blob/fce5096afdec48b54dfd504f2a897933beb640d6/src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs#L68

Not all failing tests are R2R though...

dsplaisted commented:
@wli3 From the output the It_publishes_the_project_with_a_refs_folder_and_correct_deps_file tests are failing because a new file has been added to the runtime, and the baseline needs to be updated:

> Unexpected compile libraries: System.Text.Encoding.CodePages.dll

The crossgen failures are failing with:

> /__w/1/s/artifacts/bin/Debug/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(245,5): error NETSDK1094: Unable to optimize assemblies for performance: a valid runtime package was not found. Either set the PublishReadyToRun property to false, or use a supported runtime identifier when publishing. [/__w/1/s/artifacts/tmp/Debug/It_creates_re---D21803AE/CrossgenTest3/CrossgenTest3.csproj]

That looks like it's probably also related to the runtime pack rename

dsplaisted committed:
Merge pull request #3330 from dsplaisted/fix-design-time-build-hash

Don't hash DesignTimeBould for ResolvePackageAssets

Commit:
Update stage 0

dsplaisted committed:
Update FrameworkReference tests for updated runtime pack names

dsplaisted committed:
Add System.Text.Encoding.CodePages.dll to expected self-contained output for .NET Core 3.0

dsplaisted committed:
Fix PrepareForReadyToRunCompilation task after runtime pack rename

Use FrameworkName metadata to find the runtime pack for Microsoft.NETCore.App,
instead of a now-outdated convention.

dsplaisted committed:
Merge pull request #3333 from dsplaisted/wli3-update-stage0-061419

Update stage 0

dsplaisted commented:
@vatsan-madhavan This should hopefully be unblocked if you rebase it on top of #3333, which was just merged

vatsan-madhavan committed:
Merging

vatsan-madhavan commented:
The build finally passed! 

Appreciate it if someone can merge at earliest convenience!
